server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -550,136 +562,194 @@ private Executor getExecutor(IndexShard indexShard) {
+                }
+            } else {
+                readerContext = new ReaderContext(idGenerator.incrementAndGet(), shard, reader, keepAlive,
+                    request.keepAlive() == null);
+            }
+            reader = null;
+            final ReaderContext finalReaderContext = readerContext;
+            final SearchOperationListener searchOperationListener = shard.getSearchOperationListener();
+            searchOperationListener.onNewReaderContext(finalReaderContext);
+            if (finalReaderContext.scrollContext() != null) {
+                searchOperationListener.onNewScrollContext(finalReaderContext);
+            }
+            readerContext.addOnClose(() -> {
+                try {
+                    if (finalReaderContext.scrollContext() != null) {
+                        searchOperationListener.onFreeScrollContext(finalReaderContext);
+                    }
+                } finally {
+                    searchOperationListener.onFreeReaderContext(finalReaderContext);
+                }
+            });
+            putReaderContext(finalReaderContext);
+            readerContext = null;
+            return finalReaderContext;
         } finally {
-            if (success == false) {
-                freeContext(context.id());
+            if (readerContext != null) {
+                removeReaderContext(readerContext.id().getId());
             }
+            Releasables.close(reader, readerContext, decreaseScrollContexts);
         }
     }
 
-    private void onNewContext(SearchContext context) {
-        boolean success = false;
-        try {
-            if (context.scrollContext() != null) {
-                context.indexShard().getSearchOperationListener().onNewScrollContext(context);
-            }
-            context.indexShard().getSearchOperationListener().onNewContext(context);
-            success = true;
-        } finally {
-            // currently, the concrete listener is CompositeListener, which swallows exceptions, but here we anyway try to do the
-            // right thing by closing and notifying onFreeXXX in case one of the listeners fails with an exception in the future.
-            if (success == false) {
-                try (context) {
-                    onFreeContext(context);
+    /**
+     * Opens the reader context for given shardId. The newly opened reader context will be keep
+     * until the {@code keepAlive} elapsed unless it is manually released.
+     */
+    public void openReaderContext(ShardId shardId, TimeValue keepAlive, ActionListener<SearchContextId> listener) {
+        checkKeepAliveLimit(keepAlive.millis());
+        final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        final IndexShard shard = indexService.getShard(shardId.id());
+        final SearchOperationListener searchOperationListener = shard.getSearchOperationListener();
+        shard.awaitShardSearchActive(ignored -> {
```

Comment:
I am not very clear on what thread pool we are going to execute this code after a shard has been refreshed?  Refresh thread pool?

Reply:
If the shard is not search-idle, then this will be called immediately using the network thread pool; otherwise, it will be called using the refresh thread pool.

Reply:
Thanks for clarifying, Nhat